File: php/testcode/ldapi_007.php
<?php
require_once __DIR__ . '/shared.php';

// vuln-code-snippet start php_ldapi_add_injection
function ldapi007(BenchmarkRequest $req): BenchmarkResponse {
    $conn = ldap_connect('ldap://localhost');
    $cn = $req->post('cn');
    $entry = ['cn' => $cn, 'objectClass' => 'inetOrgPerson', 'sn' => 'User'];
    ldap_add($conn, "cn=" . $cn . ",dc=example,dc=com", $entry); // vuln-code-snippet vuln-line php_ldapi_add_injection
    return BenchmarkResponse::ok('User added');
}
// vuln-code-snippet end php_ldapi_add_injection

File: rust/testcode/ldapi_007.rs
//! CWE-90: User-controlled cn injected into LDAP bind DN, enabling DN injection.

fn ldap_bind(bind_dn: &str, password: &str) -> bool {
    // In production: ldap3::LdapConn::new("ldap://...").simple_bind(bind_dn, password)
    let _ = (bind_dn, password);
    true
}

// vuln-code-snippet start testcodeLdapi007
pub fn handle(req: &super::shared::BenchmarkRequest) -> super::shared::BenchmarkResponse {
    let cn = req.param("cn");
    let password = req.param("password");
    let bind_dn = format!("cn={},ou=users,dc=example,dc=com", cn); // vuln-code-snippet target-line testcodeLdapi007
    let authenticated = ldap_bind(&bind_dn, &password);
    if authenticated {
        super::shared::BenchmarkResponse::ok("authenticated")
    } else {
        super::shared::BenchmarkResponse::forbidden("authentication failed")
    }
}
// vuln-code-snippet end testcodeLdapi007
